services/client/storage_helper.go

```diff
@@ -0,0 +1,167 @@
+
+import (
+	"errors"
+
+	"github.com/golang/protobuf/proto"
+	schema "github.com/m3db/m3cluster/generated/proto/placement"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/proto/util"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+)
+
+var (
+	errInvalidProtoForSinglePlacement    = errors.New("invalid proto for single placement")
+	errInvalidProtoForPlacementSnapshots = errors.New("invalid proto for placement snapshots")
+	errNoPlacementInTheSnapshots         = errors.New("not placement in the snapshots")
+)
+
+type placementStorageHelper interface {
+	// Placement retrieves the placement stored on kv.Store.
+	Placement(store kv.Store, key string) (services.Placement, int, error)
+
+	// Proto retrieves the proto stored on kv.Store.
+	Proto(store kv.Store, key string) (proto.Message, int, error)
+
+	// GenerateProto generates the proto message for the new placement, it may read the kv.Store
+	// if existing placement data is needed.
+	GenerateProto(store kv.Store, key string, p services.Placement) (proto.Message, error)
+
+	// ValidateProto validates if the given proto message is valid for placement.
+	ValidateProto(proto proto.Message) error
+}
+
+// newHelper returns a new placement storage helper.
+func newHelper(opts services.PlacementOptions) placementStorageHelper {
+	if opts.ShouldKeepSnapshots() {
+		return newPlacementSnapshotsHelper()
+	}
+
+	return newSinglePlacementHelper()
+}
+
+type singlePlacementHelper struct{}
+
+func newSinglePlacementHelper() placementStorageHelper {
+	return singlePlacementHelper{}
+}
+
+func (singlePlacementHelper) Placement(store kv.Store, key string) (services.Placement, int, error) {
+	v, err := store.Get(key)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	p, err := placementFromValue(v)
+	return p, v.Version(), err
+}
+
+func (singlePlacementHelper) Proto(store kv.Store, key string) (proto.Message, int, error) {
+	v, err := store.Get(key)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	p, err := placementProtoFromValue(v)
+	return p, v.Version(), err
+}
+
+func (singlePlacementHelper) GenerateProto(store kv.Store, key string, p services.Placement) (proto.Message, error) {
+	return util.PlacementToProto(p)
+}
+
+func (singlePlacementHelper) ValidateProto(proto proto.Message) error {
+	placementProto, ok := proto.(*schema.Placement)
+	if !ok {
+		return errInvalidProtoForSinglePlacement
+	}
+
+	p, err := placement.NewPlacementFromProto(placementProto)
+	if err != nil {
+		return err
+	}
+
+	return placement.Validate(p)
+}
+
+type placementSnapshotsHelper struct{}
+
+func newPlacementSnapshotsHelper() placementStorageHelper {
+	return placementSnapshotsHelper{}
+}
+
+// Placement returns the last placement in the snapshots.
+func (h placementSnapshotsHelper) Placement(store kv.Store, key string) (services.Placement, int, error) {
+	ps, v, err := h.placements(store, key)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	l := len(ps)
+	if l <= 0 {
```

Comment:
nit: `if l == 0` is probably sufficient?

Reply:
sure could change